### PR TITLE
Implement Discovery Server Example [15464]

### PR DESCRIPTION
**Pull request by jparisu**
Signed-off-by: jparisu <javierparis@eprosima.com>

## Description

Related documentation PR: eProsima/Fast-DDS-Docs#288

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- N/A Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- N/A Any new/modified methods have been properly documented using Doxygen. 
- N/A Fast DDS test suite has been run locally. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- [x] Documentation builds and tests pass locally. 
- N/A New feature has been added to the `versions.md` file (if applicable).
- N/A New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [N/A] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: examples/cpp/dds/DiscoveryServerExample/arg_configuration.h [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+        Arg::Numeric,
+        "  -p <num> \t--connection-port=<num>  \tServer listening port (Default port: 16166)."
+    },
+    {
+        TRANSPORT,
+        0,
+        "t",
+        "transport",
+        Arg::Transport,
+      "  -t <trans>\t--transport <trans> \tUse Transport Protocol [udpv4|udpv6] (UDPv6 by default)."
```
Comment (jparisu) [inline examples/cpp/dds/DiscoveryServerExample/arg_configuration.h:237]: Please, update the example to use UDPv4 by default. This implies:
- Change in main the default arguments
- Change this help messages

**Review by jparisu — COMMENTED** [review]
Requested changes

File: examples/cpp/dds/DiscoveryServerExample/arg_configuration.h (hunk 1 above)
Comment (jparisu) [reply examples/cpp/dds/DiscoveryServerExample/arg_configuration.h]: It must change also de default value (and its messages) for the listening and connection address:
`::1` -> `127.0.0.1`

File: versions.md [hunk 2]
```diff
@@ -1,6 +1,11 @@
 Forthcoming
 -----------
 
+Version 2.10.0
+-------------
+
+* Implement Discovery Server example
+
```
Comment (EduPonz) [inline versions.md:8]: Normally we do not include examples in the versions.md. I think you can just mark that check as *N/A*

```suggestion
```

**Review by EduPonz — CHANGES_REQUESTED** [review]

**Comment by JLBuenoLopez-eProsima** [comment]
@jparisu @irenebm Does this PR supersedes #2178? If so, can you close it?

**Comment by juanlofer-eprosima** [comment]
After merging this PR, add reference to documentation (https://github.com/eProsima/Fast-DDS-docs/pull/288).

**Comment by JLBuenoLopez-eProsima** [comment]
@richiprosima please test this

File: examples/cpp/dds/DiscoveryServerExample/README.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+In the first one launch: ./DiscoveryServerExample publisher (or DiscoveryServerExample.exe publisher on windows).
+In the second one: ./DiscoveryServerExample subscriber (or DiscoveryServerExample.exe subscriber on windows).
+In the third one: ./DiscoveryServerExample server (or DiscoveryServerExample.exe server on windows).
+
+## Arguments
+
+First argument is `publisher`, `subscriber` or `server` and then the rest of arguments are read unordered
+
+```sh
+Usage: DiscoveryServerExample <publisher|subscriber|server>
```
Comment (jparisu) [inline examples/cpp/dds/DiscoveryServerExample/README.md:19]: Update

File: examples/cpp/dds/DiscoveryServerExample/DiscoveryServer_main.cpp [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+                    else
+                    {
+                        print_warning("udpv4|udpv6|shm", opt.name);
+                    }
+
+                    break;
+                }
+
+                case optionIndex::CONNECTION_PORT:
+                    connection_port = strtol(opt.arg, nullptr, 10);
```
Comment (jparisu) [inline examples/cpp/dds/DiscoveryServerExample/DiscoveryServer_main.cpp:189]: `'=': conversion from 'long' to 'uint16_t', possible loss of data`

File: examples/cpp/dds/DiscoveryServerExample/DiscoveryServer_main.cpp [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+                    break;
+
+                case optionIndex::CONNECTION_ADDRESS:
+                    connection_address = opt.arg;
+                    break;
+
+                case optionIndex::CONNECTION_DISCOVERY_SERVER_ID:
+                    id_ds_set = true;
+                    connection_ds_id = strtol(opt.arg, nullptr, 10);
```
Comment (jparisu) [inline examples/cpp/dds/DiscoveryServerExample/DiscoveryServer_main.cpp:198]: `'=': conversion from 'long' to 'uint16_t', possible loss of data`

File: examples/cpp/dds/DiscoveryServerExample/DiscoveryServer_main.cpp [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+                    break;
+
+                case optionIndex::LISTENING_PORT:
+                    if (type != EntityKind::SERVER)
+                    {
+                        print_warning("server", opt.name);
+                        break;
+                    }
+                    listening_port = strtol(opt.arg, nullptr, 10);
```
Comment (jparisu) [inline examples/cpp/dds/DiscoveryServerExample/DiscoveryServer_main.cpp:207]: `'=': conversion from 'long' to 'uint16_t', possible loss of data`

File: examples/cpp/dds/DiscoveryServerExample/DiscoveryServer_main.cpp [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+
+                    break;
+
+                case optionIndex::LISTENING_DISCOVERY_SERVER_ID:
+                    if (type != EntityKind::SERVER)
+                    {
+                        print_warning("server", opt.name);
+                        break;
+                    }
+                    listening_ds_id = strtol(opt.arg, nullptr, 10);
```
Comment (jparisu) [inline examples/cpp/dds/DiscoveryServerExample/DiscoveryServer_main.cpp:226]: `'=': conversion from 'long' to 'uint16_t', possible loss of data`

**Review by jparisu — COMMENTED** [review]
@irenebm please apply changes

**Comment by JesusPoderoso** [comment]
@richiprosima please test this

File: examples/cpp/dds/DiscoveryServerExample/DiscoveryServer_main.cpp [hunk 8]
```diff
@@ -186,7 +186,7 @@ int main(
                 }
 
                 case optionIndex::CONNECTION_PORT:
-                    connection_port = strtol(opt.arg, nullptr, 10);
+                    connection_port = static_cast<uint16_t>strtol(opt.arg, nullptr, 10);
```
Comment (jparisu) [inline examples/cpp/dds/DiscoveryServerExample/DiscoveryServer_main.cpp:189]:
```
connection_port = static_cast<uint16_t>(strtol(opt.arg, nullptr, 10));
```
Please, compile in local before upload a commit

Same error is repeated in other casts

**Review by jparisu — COMMENTED** [review]
@irenebm please check this

File: examples/cpp/dds/DiscoveryServerExample/README.md [hunk 9]
```diff
@@ -29,22 +29,50 @@ Publisher options:
                     Number of samples to send (Default: 0 => infinite samples).
   -i <num>        --interval=<num>
                     Time between samples in milliseconds (Default: 100).
-                  --connection-address=<IPaddress[:port number]>
-                    Server address (Default address: 127.0.0.1, default port:
-                    60006).
+  -c <IPaddress>  --connection-address=<IPaddress[:port number]>
```
[the remark this answers is not in the document]
Comment (jparisu) [reply examples/cpp/dds/DiscoveryServerExample/README.md]: Not correctly done. This readme is finally a copy of the args set in `arg_configuration.h`

**Review by irenebm — APPROVED** [review]
LGTM

**Comment by jparisu** [comment]
@richiprosima please test this

**Pull request merged**